api/peer/list.go

```diff
@@ -59,6 +53,26 @@ type ListUpdates struct {
 	Removals []Identifier
 }
 
+// ChooserList is both a Chooser and a List, useful for expressing both
+// capabilities of a single instance.
+type ChooserList interface {
+	Chooser
+	List
+}
+
+// RetainedChooserList is a collection of retained peers, with its own
+// subscribers for peer status change notifications.
+// The retained peer chooser list encapsulates the logic for selecting from
+// among available peers, whereas a ChooserList is responsible for retaining,
+// releasing, and monitoring peer availability.
+// Use "go.uber.org/yarpc/peer/peerlist".List in conjunction with a
+// RetainedList to produce a "go.uber.org/yarpc/api/peer".List.
+type RetainedChooserList interface {
+	Add(Peer) Subscriber
+	Remove(Peer, Subscriber)
+	Choose(context.Context, *transport.Request) Peer
```

Comment:
Is there a possibility that the Chooser should be able to reject a "Choose" request?  I'm thinking of sharding cases where a request for a shard that is not valid is made.  Or if the List is empty.

Reply:
In the case that the list is empty, each of the implementations returns nil, and peerlist cycles back until a peer becomes available again. We could make this (Peer, error) to be surprise-proof, and only loop around if the error is a YARPC unavailable error, but there is a danger that this would make a tight loop, since the peerlist’s contract is to only put available peers in the retained chooser list.

Relatedly, I’m inclined to also thread transport.Lifecycle just in case we need it (or punt lifecycle to dig/fx at this point).

Reply:
Should we just have the RetainedChooserList implement the Chooser API?  Might add some cruft, but it moves a bit closer to the "middleware" area, for Choosers that don't need the extra error/func(err) we can add a wrapper that will convert between this API and the Chooser API version.  

Definitely not a blocker since this API surface is very flexible and we can wrap as much as we want with no consequences :)